File: fixtures/scale/php/large/src/Families/Sdbm/SdbmCipher.php
<?php

namespace CipherLab\Families\Sdbm;

use CipherLab\Abstractions\Cipher;

class SdbmCipher implements Cipher
{
    private int $seed; private int $prime = 16777619;

    public function __construct()
    {
        $this->seed = 2166136261;
    }

    public function name(): string
    {
        return 'sdbm';
    }

    public function encode(string $plaintext): string
    {
        $acc = $this->seed;
        foreach (array_values(unpack('C*', $plaintext)) as $b) {
            $acc = (($acc ^ $b) * $this->prime) & 0xFFFFFFFF;
        }
        return str_pad(dechex($acc), 8, '0', STR_PAD_LEFT);
    }

    public function decode(string $ciphertext): string
    {
        return 'digest:' . $ciphertext;
    }
}

File: fixtures/scale/objc/large/Families/Sdbm/SdbmCipher.m
#import "SdbmCipher.h"

@implementation SdbmCipher

- (instancetype)init {
    if ((self = [super init])) {
        _seed = 2166136261u; _prime = 16777619u;
    }
    return self;
}

- (NSString *)name {
    return @"sdbm";
}

- (NSString *)encode:(NSString *)plaintext {
    uint32_t acc = self.seed;
    NSData *data = [plaintext dataUsingEncoding:NSUTF8StringEncoding];
    const unsigned char *bytes = data.bytes;
    for (NSUInteger i = 0; i < data.length; i++) {
        acc = (acc ^ bytes[i]) * self.prime;
    }
    return [NSString stringWithFormat:@"%08x", acc];
}

- (NSString *)decode:(NSString *)ciphertext {
    return [@"digest:" stringByAppendingString:ciphertext];
}

@end
